// react-client/dist/server/broadcastManager.js
import { WebSocketServer, WebSocket as NodeWebSocket } from 'ws';
import chalk from 'chalk';
/**
 * BroadcastManager — Shared WebSocket utility for dev, preview, and SSR servers.
 * Generic over message type T which defaults to HMRMessage.
 */
export class BroadcastManager {
    wss;
    clients = new Set();
    constructor(server) {
        this.wss = new WebSocketServer({ server });
        this.wss.on('connection', (ws) => {
            this.clients.add(ws);
            ws.on('close', () => {
                this.clients.delete(ws);
            });
            ws.on('error', (err) => {
                console.error(chalk.red('⚠️ WebSocket error:'), err.message);
            });
        });
    }
    broadcast(msg) {
        const data = JSON.stringify(msg);
        for (const client of this.clients) {
            if (client.readyState === NodeWebSocket.OPEN) {
                client.send(data);
            }
        }
    }
    send(ws, msg) {
        if (ws.readyState === NodeWebSocket.OPEN) {
            ws.send(JSON.stringify(msg));
        }
    }
    getClientCount() {
        return this.clients.size;
    }
    close() {
        console.log(chalk.red('🛑 Closing WebSocket connections...'));
        this.wss.close();
        for (const ws of this.clients) {
            try {
                ws.close();
            }
            catch {
                // ignore
            }
        }
        this.clients.clear();
    }
}
